File: web/lib/shell/run-summary.ts
/** The one persisted recipe-run summary accepted by the web consumer. */

export const RECIPE_RUN_SCHEMA_VERSION = 3 as const;
export const RECIPE_RUN_KIND = "recipe_run_v3" as const;

export type RunTransparencyMode = "ai" | "chroma";

export interface JsonObject {
  readonly [key: string]: JsonValue;
}

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | readonly JsonValue[]
  | JsonObject;

export type SuccessfulRecipeRunStage = Readonly<{
  stage: string;
  ok: true;
  duration_ms: number;
  artifacts: readonly string[];
}>;

export type FailedRecipeRunStage = Readonly<{
  stage: string;
  ok: false;
  duration_ms: number;
  artifacts: readonly string[];
  error: string;
}>;

export type RecipeRunStage = SuccessfulRecipeRunStage | FailedRecipeRunStage;

type RecipeRunSummaryFields = Readonly<{
  schema_version: typeof RECIPE_RUN_SCHEMA_VERSION;
  kind: typeof RECIPE_RUN_KIND;
  recipe: string;
  input: Readonly<Record<string, JsonValue>> & {
    readonly transparency_mode: RunTransparencyMode;
  };
  tag: string;
  run_dir: string;
  started_at: string;
  ended_at: string;
  duration_ms: number;
}>;

export type SuccessfulRecipeRunSummary = RecipeRunSummaryFields &
  Readonly<{
    ok: true;
    stages: readonly SuccessfulRecipeRunStage[];
  }>;

export type FailedRecipeRunSummary = RecipeRunSummaryFields &
  Readonly<{
    ok: false;
    stages: readonly RecipeRunStage[];
    failed_stage: string;
  }>;

export type RecipeRunSummary = SuccessfulRecipeRunSummary | FailedRecipeRunSummary;

export type RunCompletionPayload = Readonly<{
  ok: boolean;
  failed_stage: string | null;
}>;

const SUCCESS_RUN_KEYS = [
  "schema_version",
  "kind",
  "recipe",
  "input",
  "tag",
  "run_dir",
  "started_at",
  "ended_at",
  "duration_ms",
  "ok",
  "stages",
] as const;
const FAILED_RUN_KEYS = [...SUCCESS_RUN_KEYS, "failed_stage"] as const;
const SUCCESS_STAGE_KEYS = ["stage", "ok", "duration_ms", "artifacts"] as const;
const FAILED_STAGE_KEYS = [...SUCCESS_STAGE_KEYS, "error"] as const;
const STABLE_ID = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const LOWER_SNAKE_FIELD = /^[a-z][a-z0-9]*(?:_[a-z0-9]+)*$/;
const SAFE_TAG = /^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$/;
const ARTIFACT_SEGMENT = /^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$/;
const UTC_TIMESTAMP = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.(\d{1,6}))?Z$/;

function fail(path: string, message: string): never {
  throw new Error(`${path} ${message}`);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

function record(value: unknown, path: string): Record<string, unknown> {
  if (!isRecord(value)) return fail(path, "must be an object");
  return value;
}

function exactKeys(
  value: Record<string, unknown>,
  expected: readonly string[],
  path: string,
): void {
  const keys = new Set(expected);
  for (const key of Object.keys(value)) {
    if (!keys.has(key)) fail(`${path}.${key}`, "is not a supported key");
  }
  for (const key of expected) {
    if (!Object.prototype.hasOwnProperty.call(value, key)) {
      fail(`${path}.${key}`, "is required");
    }
  }
}

function literal<const Value extends string | number>(
  value: unknown,
  expected: Value,
  path: string,
): Value {
  if (value !== expected) return fail(path, `must equal ${JSON.stringify(expected)}`);
  return expected;
}

function bool(value: unknown, path: string): boolean {
  if (typeof value !== "boolean") return fail(path, "must be a boolean");
  return value;
}

function stableText(value: unknown, path: string): string {
  if (
    typeof value !== "string" ||
    !value ||
    value !== value.trim() ||
    value.includes("\0")
  ) {
    return fail(path, "must be a non-empty trimmed string");
  }
  return value;
}

function stableId(value: unknown, path: string): string {
  const parsed = stableText(value, path);
  if (!STABLE_ID.test(parsed)) {
    return fail(path, "must be a lower-case hyphenated identifier");
  }
  return parsed;
}

function safeTag(value: unknown, path: string): string {
  const parsed = stableText(value, path);
  if (!SAFE_TAG.test(parsed) || parsed === "." || parsed === "..") {
    return fail(path, "must be one safe path segment");
  }
  return parsed;
}

function nonNegativeInteger(value: unknown, path: string): number {
  if (!Number.isSafeInteger(value) || (value as number) < 0) {
    return fail(path, "must be a non-negative safe integer");
  }
  return value as number;
}

type ParsedTimestamp = Readonly<{
  value: string;
  parts: readonly [number, number, number, number, number, number, string];
}>;

function timestamp(value: unknown, path: string): ParsedTimestamp {
  const parsed = stableText(value, path);
  const match = UTC_TIMESTAMP.exec(parsed);
  if (!match) {
    return fail(path, "must be a valid UTC timestamp ending in Z");
  }
  const parts = [
    Number(parsed.slice(0, 4)),
    Number(parsed.slice(5, 7)),
    Number(parsed.slice(8, 10)),
    Number(parsed.slice(11, 13)),
    Number(parsed.slice(14, 16)),
    Number(parsed.slice(17, 19)),
    match[1] ?? "",
  ] as const;
  const [year, month, day, hour, minute, second] = parts;
  const leapYear = year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
  const monthLengths = [31, leapYear ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  if (
    year < 1 ||
    month < 1 ||
    month > 12 ||
    day < 1 ||
    day > monthLengths[month - 1] ||
    hour > 23 ||
    minute > 59 ||
    second > 59
  ) {
    return fail(path, "must be a valid UTC timestamp ending in Z");
  }
  return { value: parsed, parts };
}

function timestampPrecedes(left: ParsedTimestamp, right: ParsedTimestamp): boolean {
  for (let index = 0; index < 6; index += 1) {
    const leftPart = left.parts[index] as number;
    const rightPart = right.parts[index] as number;
    if (leftPart !== rightPart) return leftPart < rightPart;
  }
  const leftFraction = left.parts[6];
  const rightFraction = right.parts[6];
  const width = Math.max(leftFraction.length, rightFraction.length);
  return leftFraction.padEnd(width, "0") < rightFraction.padEnd(width, "0");
}

function jsonValue(value: unknown, path: string): JsonValue {
  if (value === null || typeof value === "boolean" || typeof value === "string") return value;
  if (typeof value === "number") {
    if (!Number.isFinite(value)) return fail(path, "must be finite JSON data");
    if (Number.isInteger(value) && !Number.isSafeInteger(value)) {
      return fail(path, "must keep integers within the JSON safe-integer range");
    }
    return value;
  }
  if (Array.isArray(value)) {
    return Object.freeze(
      Array.from(value, (entry, index) => jsonValue(entry, `${path}[${index}]`)),
    );
  }
  if (!isRecord(value)) return fail(path, "must be JSON data");
  return Object.freeze(
    Object.fromEntries(
      Object.entries(value).map(([key, entry]) => [
        key,
        jsonValue(entry, `${path}.${key}`),
      ]),
    ),
  );
}

function runInput(value: unknown): RecipeRunSummary["input"] {
  const root = record(value, "run_summary.input");
  if (Object.prototype.hasOwnProperty.call(root, "transparencyMode")) {
    fail(
      "run_summary.input.transparencyMode",
      "is not supported; use transparency_mode",
    );
  }
  for (const key of Object.keys(root)) {
    if (!LOWER_SNAKE_FIELD.test(key)) {
      fail(`run_summary.input.${key}`, "must use lower_snake_case");
    }
  }
  const mode = root["transparency_mode"];
  if (mode !== "ai" && mode !== "chroma") {
    fail("run_summary.input.transparency_mode", "must be ai or chroma");
  }
  return jsonValue(root, "run_summary.input") as RecipeRunSummary["input"];
}

function portableArtifactPath(value: unknown, path: string): string {
  const parsed = stableText(value, path);
  const segments = parsed.split("/");
  if (
    parsed.includes("\0") ||
    parsed.startsWith("/") ||
    segments.some((segment) => !ARTIFACT_SEGMENT.test(segment))
  ) {
    return fail(path, "must be a portable relative POSIX path");
  }
  return parsed;
}

function artifacts(value: unknown, path: string): readonly string[] {
  if (!Array.isArray(value)) return fail(path, "must be an array");
  const parsed = Object.freeze(
    Array.from(value, (entry, index) => portableArtifactPath(entry, `${path}[${index}]`)),
  );
  if (new Set(parsed).size !== parsed.length) {
    fail(path, "must not contain duplicate paths");
  }
  return parsed;
}

function runStage(value: unknown, index: number): RecipeRunStage {
  const path = `run_summary.stages[${index}]`;
  const root = record(value, path);
  const ok = bool(root["ok"], `${path}.ok`);
  exactKeys(root, ok ? SUCCESS_STAGE_KEYS : FAILED_STAGE_KEYS, path);
  const stage = stableId(root["stage"], `${path}.stage`);
  const durationMs = nonNegativeInteger(root["duration_ms"], `${path}.duration_ms`);
  const stageArtifacts = artifacts(root["artifacts"], `${path}.artifacts`);
  if (ok) {
    return Object.freeze({ stage, ok: true, duration_ms: durationMs, artifacts: stageArtifacts });
  }
  if (stageArtifacts.length !== 0) {
    fail(`${path}.artifacts`, "must be empty for a failed stage");
  }
  const error = stableText(root["error"], `${path}.error`);
  return Object.freeze({
    stage,
    ok: false,
    duration_ms: durationMs,
    artifacts: Object.freeze([]),
    error,
  });
}

/** Parse and validate the exact current persisted run-summary contract. */
export function parseRecipeRunSummary(value: unknown): RecipeRunSummary {
  const root = record(value, "run_summary");
  const ok = bool(root["ok"], "run_summary.ok");
  exactKeys(root, ok ? SUCCESS_RUN_KEYS : FAILED_RUN_KEYS, "run_summary");
  const schemaVersion = literal(
    root["schema_version"],
    RECIPE_RUN_SCHEMA_VERSION,
    "run_summary.schema_version",
  );
  const kind = literal(root["kind"], RECIPE_RUN_KIND, "run_summary.kind");
  const recipe = stableId(root["recipe"], "run_summary.recipe");
  const input = runInput(root["input"]);
  const tag = safeTag(root["tag"], "run_summary.tag");
  const runDir = safeTag(root["run_dir"], "run_summary.run_dir");
  if (runDir !== tag) {
    fail("run_summary.run_dir", "must equal run_summary.tag");
  }
  const startedAt = timestamp(root["started_at"], "run_summary.started_at");
  const endedAt = timestamp(root["ended_at"], "run_summary.ended_at");
  if (timestampPrecedes(endedAt, startedAt)) {
    fail("run_summary.ended_at", "must not precede started_at");
  }
  const durationMs = nonNegativeInteger(root["duration_ms"], "run_summary.duration_ms");
  if (!Array.isArray(root["stages"])) {
    fail("run_summary.stages", "must be an array");
  }
  if (root["stages"].length === 0) {
    fail("run_summary.stages", "must contain at least one executed stage");
  }
  const stages = Object.freeze(
    Array.from(root["stages"], (stage, index) => runStage(stage, index)),
  );
  const stageIds = stages.map((stage) => stage.stage);
  if (new Set(stageIds).size !== stageIds.length) {
    fail("run_summary.stages", "must contain unique stage identifiers");
  }
  const failed = stages.filter((stage): stage is FailedRecipeRunStage => !stage.ok);

  const common = {
    schema_version: schemaVersion,
    kind,
    recipe,
    input,
    tag,
    run_dir: runDir,
    started_at: startedAt.value,
    ended_at: endedAt.value,
    duration_ms: durationMs,
  } as const;

  if (ok) {
    if (failed.length !== 0) {
      fail("run_summary.stages", "must all succeed when run_summary.ok is true");
    }
    return Object.freeze({
      ...common,
      ok: true,
      stages: stages as readonly SuccessfulRecipeRunStage[],
    });
  }

  if (failed.length !== 1 || stages.at(-1) !== failed[0]) {
    fail("run_summary.stages", "must end with exactly one failed stage when run_summary.ok is false");
  }
  const failedStage = stableId(root["failed_stage"], "run_summary.failed_stage");
  if (failedStage !== failed[0].stage) {
    fail("run_summary.failed_stage", "must match the final failed stage");
  }
  return Object.freeze({
    ...common,
    ok: false,
    stages,
    failed_stage: failedStage,
  });
}

class DuplicateJsonKeyError extends Error {}

/** JSON.parse discards duplicate object keys, so scan the valid source before accepting it. */
function rejectDuplicateJsonKeys(source: string): void {
  let cursor = 0;

  const skipWhitespace = () => {
    while (
      source[cursor] === " " ||
      source[cursor] === "\t" ||
      source[cursor] === "\n" ||
      source[cursor] === "\r"
    ) {
      cursor += 1;
    }
  };

  const readString = (): string => {
    const start = cursor;
    cursor += 1;
    while (cursor < source.length) {
      const character = source[cursor];
      if (character === '"') {
        cursor += 1;
        return JSON.parse(source.slice(start, cursor)) as string;
      }
      if (character === "\\") {
        cursor += 1;
        cursor += source[cursor] === "u" ? 5 : 1;
      } else {
        cursor += 1;
      }
    }
    throw new SyntaxError("unterminated JSON string");
  };

  const readValue = (): void => {
    skipWhitespace();
    const character = source[cursor];
    if (character === "{") {
      readObject();
      return;
    }
    if (character === "[") {
      readArray();
      return;
    }
    if (character === '"') {
      readString();
      return;
    }
    for (const literalValue of ["true", "false", "null"] as const) {
      if (source.startsWith(literalValue, cursor)) {
        cursor += literalValue.length;
        return;
      }
    }
    const number = source.slice(cursor).match(/^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?/);
    if (!number) throw new SyntaxError("invalid JSON value");
    cursor += number[0].length;
  };

  const readObject = (): void => {
    cursor += 1;
    skipWhitespace();
    const keys = new Set<string>();
    if (source[cursor] === "}") {
      cursor += 1;
      return;
    }
    while (cursor < source.length) {
      skipWhitespace();
      if (source[cursor] !== '"') throw new SyntaxError("invalid JSON object key");
      const key = readString();
      if (keys.has(key)) throw new DuplicateJsonKeyError(`duplicate JSON key: ${key}`);
      keys.add(key);
      skipWhitespace();
      if (source[cursor] !== ":") throw new SyntaxError("invalid JSON object entry");
      cursor += 1;
      readValue();
      skipWhitespace();
      if (source[cursor] === "}") {
        cursor += 1;
        return;
      }
      if (source[cursor] !== ",") throw new SyntaxError("invalid JSON object separator");
      cursor += 1;
    }
    throw new SyntaxError("unterminated JSON object");
  };

  const readArray = (): void => {
    cursor += 1;
    skipWhitespace();
    if (source[cursor] === "]") {
      cursor += 1;
      return;
    }
    while (cursor < source.length) {
      readValue();
      skipWhitespace();
      if (source[cursor] === "]") {
        cursor += 1;
        return;
      }
      if (source[cursor] !== ",") throw new SyntaxError("invalid JSON array separator");
      cursor += 1;
    }
    throw new SyntaxError("unterminated JSON array");
  };

  readValue();
  skipWhitespace();
  if (cursor !== source.length) throw new SyntaxError("trailing JSON data");
}

export function parseRecipeRunSummaryText(source: string): RecipeRunSummary {
  let value: unknown;
  try {
    value = JSON.parse(source);
    rejectDuplicateJsonKeys(source);
  } catch (error) {
    if (error instanceof DuplicateJsonKeyError) {
      throw new Error(`run summary is not valid JSON: ${error.message}`);
    }
    throw new Error("run summary must be valid JSON");
  }
  return parseRecipeRunSummary(value);
}

export function parseRecipeRunSummaryBytes(source: Uint8Array): RecipeRunSummary {
  if (source[0] === 0xef && source[1] === 0xbb && source[2] === 0xbf) {
    throw new Error("run summary must be BOM-free UTF-8 JSON");
  }
  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(source);
  } catch {
    throw new Error("run summary must be valid UTF-8 JSON");
  }
  return parseRecipeRunSummaryText(text);
}

/** Explicit lower_snake_case projection used by the public pipeline-done SSE event. */
export function runCompletionPayload(summary: RecipeRunSummary): RunCompletionPayload {
  return Object.freeze({
    ok: summary.ok,
    failed_stage: summary.ok ? null : summary.failed_stage,
  });
}

File: src/stage_gen/contracts/run_summary.py
"""Exact current persisted recipe-run summary contract."""

from __future__ import annotations

import json
import math
import re
from datetime import datetime
from pathlib import Path
from typing import Any, Literal, Self

from pydantic import Field, ValidationError, field_validator, model_validator

from gnode import PersistedContractModel

RUN_SUMMARY_SCHEMA_VERSION = 3
RUN_SUMMARY_KIND = "recipe_run_v3"
MAX_JSON_SAFE_INTEGER = 9_007_199_254_740_991

_IDENTIFIER = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
_LOWER_SNAKE_FIELD = re.compile(r"^[a-z][a-z0-9]*(?:_[a-z0-9]+)*$")
_SAFE_TAG = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
_ARTIFACT_SEGMENT = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
_UTC_TIMESTAMP = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{1,6})?Z$")


class RecipeRunStage(PersistedContractModel):
    """One executed stage; stages skipped after failure are not serialized."""

    stage: str
    ok: bool
    duration_ms: int = Field(ge=0, le=MAX_JSON_SAFE_INTEGER)
    artifacts: list[str]
    error: str | None = None

    @field_validator("duration_ms", mode="before")
    @classmethod
    def normalize_duration(cls, value: object) -> object:
        return _normalize_semantic_integer(value)

    @field_validator("stage")
    @classmethod
    def validate_stage(cls, value: str) -> str:
        if _IDENTIFIER.fullmatch(value) is None:
            raise ValueError("stage must be a lower-kebab-case identifier")
        return value

    @field_validator("artifacts")
    @classmethod
    def validate_artifacts(cls, value: list[str]) -> list[str]:
        for item in value:
            validate_run_artifact_ref(item)
        if len(value) != len(set(value)):
            raise ValueError("stage artifacts must be unique")
        return value

    @field_validator("error")
    @classmethod
    def validate_error(cls, value: str | None) -> str | None:
        if value is not None and (not value or value != value.strip()):
            raise ValueError("stage error must be a non-empty trimmed string")
        if value is not None and "\x00" in value:
            raise ValueError("stage error must not contain NUL")
        return value

    @model_validator(mode="after")
    def validate_outcome(self) -> Self:
        error_was_serialized = "error" in self.model_fields_set
        if self.ok:
            if error_was_serialized:
                raise ValueError("successful stage must omit error")
            return self
        if not error_was_serialized or self.error is None:
            raise ValueError("failed stage must declare error")
        if self.artifacts:
            raise ValueError("failed stage must not publish artifacts")
        return self


class RecipeRunSummary(PersistedContractModel):
    """Strict, closed recipe-run document written to ``run.json``."""

    schema_version: Literal[3]
    kind: Literal["recipe_run_v3"]
    recipe: str
    input: dict[str, Any]
    tag: str
    run_dir: str
    started_at: str
    ended_at: str
    duration_ms: int = Field(ge=0, le=MAX_JSON_SAFE_INTEGER)
    ok: bool
    stages: list[RecipeRunStage] = Field(min_length=1)
    failed_stage: str | None = None

    @field_validator("duration_ms", mode="before")
    @classmethod
    def normalize_duration(cls, value: object) -> object:
        return _normalize_semantic_integer(value)

    @field_validator("recipe")
    @classmethod
    def validate_recipe(cls, value: str) -> str:
        if _IDENTIFIER.fullmatch(value) is None:
            raise ValueError("recipe must be a lower-kebab-case identifier")
        return value

    @field_validator("tag")
    @classmethod
    def validate_tag(cls, value: str) -> str:
        if _SAFE_TAG.fullmatch(value) is None or value in {".", ".."}:
            raise ValueError("tag must be one safe path segment")
        return value

    @field_validator("run_dir")
    @classmethod
    def validate_run_dir(cls, value: str) -> str:
        if _SAFE_TAG.fullmatch(value) is None or value in {".", ".."}:
            raise ValueError("run_dir must be one portable safe path segment")
        return value

    @field_validator("started_at", "ended_at")
    @classmethod
    def validate_timestamp(cls, value: str) -> str:
        if _UTC_TIMESTAMP.fullmatch(value) is None:
            raise ValueError(
                "run timestamp must be UTC ISO-8601 ending in Z with at most six fractional digits"
            )
        try:
            datetime.fromisoformat(value[:-1] + "+00:00")
        except ValueError as error:
            raise ValueError("run timestamp must be a valid UTC ISO-8601 timestamp") from error
        return value

    @field_validator("input")
    @classmethod
    def validate_input_json(cls, value: dict[str, Any]) -> dict[str, Any]:
        _assert_json_value(value, "input")
        if "transparencyMode" in value:
            raise ValueError("input must not declare legacy transparencyMode")
        invalid_fields = sorted(key for key in value if _LOWER_SNAKE_FIELD.fullmatch(key) is None)
        if invalid_fields:
            raise ValueError("input fields must use lower_snake_case: " + ", ".join(invalid_fields))
        if value.get("transparency_mode") not in {"native", "ai", "chroma"}:
            raise ValueError("input.transparency_mode must be native, ai, or chroma")
        return value

    @field_validator("failed_stage")
    @classmethod
    def validate_failed_stage(cls, value: str | None) -> str | None:
        if value is not None and _IDENTIFIER.fullmatch(value) is None:
            raise ValueError("failed_stage must be a lower-kebab-case identifier")
        return value

    @model_validator(mode="after")
    def validate_run_outcome(self) -> Self:
        if self.run_dir != self.tag:
            raise ValueError("run_dir must equal tag")
        started = datetime.fromisoformat(self.started_at[:-1] + "+00:00")
        ended = datetime.fromisoformat(self.ended_at[:-1] + "+00:00")
        if ended < started:
            raise ValueError("ended_at must not precede started_at")

        stage_names = [stage.stage for stage in self.stages]
        if len(stage_names) != len(set(stage_names)):
            raise ValueError("run stages must have unique stage identifiers")

        failed = [stage for stage in self.stages if not stage.ok]
        failed_stage_was_serialized = "failed_stage" in self.model_fields_set
        if self.ok:
            if failed_stage_was_serialized:
                raise ValueError("successful run must omit failed_stage")
            if failed:
                raise ValueError("successful run must contain only successful stages")
            return self

        if not failed_stage_was_serialized or self.failed_stage is None:
            raise ValueError("failed run must declare failed_stage")
        if len(failed) != 1:
            raise ValueError("failed run must contain exactly one failed stage")
        if not self.stages or self.stages[-1].stage != self.failed_stage:
            raise ValueError("failed_stage must identify the final executed stage")
        if failed[0].stage != self.failed_stage:
            raise ValueError("failed_stage must identify the failed stage")
        return self

    def to_dict(self) -> dict[str, Any]:
        """Return the exact persisted shape, omitting outcome-only optional fields."""

        return self.model_dump(mode="json", exclude_none=True)


class RecipeRunSummaryLoadError(ValueError):
    """Raised when a run summary is not the exact current contract."""


class _DuplicateKeyError(ValueError):
    pass


def parse_recipe_run_summary(
    value: object,
    *,
    label: str = "run summary",
) -> RecipeRunSummary:
    """Validate one decoded value as the exact current run-summary contract."""

    try:
        return RecipeRunSummary.model_validate(value)
    except ValidationError as error:
        raise RecipeRunSummaryLoadError(
            f"{label} is not a valid {RUN_SUMMARY_KIND}: {error}"
        ) from error


def load_recipe_run_summary(
    path: str | Path,
    *,
    label: str = "run summary",
) -> RecipeRunSummary:
    """Read UTF-8 JSON, reject ambiguous syntax, and validate the current contract."""

    try:
        text = Path(path).read_text(encoding="utf-8")
        value = json.loads(
            text,
            object_pairs_hook=_reject_duplicate_keys,
            parse_constant=_reject_json_constant,
        )
    except (OSError, UnicodeError, ValueError) as error:
        raise RecipeRunSummaryLoadError(f"{label} is not valid JSON: {error}") from error
    return parse_recipe_run_summary(value, label=label)


def _reject_duplicate_keys(pairs: list[tuple[str, object]]) -> dict[str, object]:
    result: dict[str, object] = {}
    for key, value in pairs:
        if key in result:
            raise _DuplicateKeyError(f"duplicate JSON key: {key}")
        result[key] = value
    return result


def _reject_json_constant(value: str) -> None:
    raise ValueError(f"invalid JSON constant: {value}")


def _assert_json_value(value: object, label: str) -> None:
    if value is None or isinstance(value, (bool, str)):
        return
    if isinstance(value, int):
        if not -MAX_JSON_SAFE_INTEGER <= value <= MAX_JSON_SAFE_INTEGER:
            raise ValueError(f"{label} contains an integer outside the JSON safe range")
        return
    if isinstance(value, float):
        if not math.isfinite(value):
            raise ValueError(f"{label} contains a non-finite number")
        if value.is_integer() and abs(value) > MAX_JSON_SAFE_INTEGER:
            raise ValueError(f"{label} contains an integer outside the JSON safe range")
        return
    if isinstance(value, list):
        for index, item in enumerate(value):
            _assert_json_value(item, f"{label}[{index}]")
        return
    if isinstance(value, dict):
        for key, item in value.items():
            if not isinstance(key, str):
                raise ValueError(f"{label} contains a non-string object key")
            _assert_json_value(item, f"{label}.{key}")
        return
    raise ValueError(f"{label} contains a non-JSON value")


def validate_run_artifact_ref(value: str) -> str:
    """Validate one run-local artifact reference as portable relative POSIX syntax."""

    if not value or value != value.strip() or "\x00" in value:
        raise ValueError("stage artifact must be a non-empty trimmed portable reference")
    if value.startswith(("/", "~")) or any(mark in value for mark in ("\\", ":", "?", "#")):
        raise ValueError("stage artifact must be a portable relative POSIX reference")
    segments = value.split("/")
    if any(_ARTIFACT_SEGMENT.fullmatch(segment) is None for segment in segments):
        raise ValueError("stage artifact must be a portable relative POSIX reference")
    return value


def _normalize_semantic_integer(value: object) -> object:
    if isinstance(value, float) and math.isfinite(value) and value.is_integer():
        return int(value)
    return value


__all__ = [
    "RUN_SUMMARY_KIND",
    "RUN_SUMMARY_SCHEMA_VERSION",
    "MAX_JSON_SAFE_INTEGER",
    "RecipeRunStage",
    "RecipeRunSummary",
    "RecipeRunSummaryLoadError",
    "load_recipe_run_summary",
    "parse_recipe_run_summary",
    "validate_run_artifact_ref",
]
